src/server/bitops_family.cc

```diff
@@ -0,0 +1,304 @@
+}
+
+uint8_t GetByteValue(const std::string& str, uint32_t offset) {
+  return static_cast<uint8_t>(str[GetByteIndex(offset)]);
+}
+
+constexpr bool CheckBitStatus(uint8_t byte, uint32_t offset) {
+  return byte & (0x1 << offset);
+}
+
+// return true if bit is on
+bool GetBitValue(const std::string& entry, uint32_t offset) {
+  const auto byte_val{GetByteValue(entry, offset)};
+  const auto index{GetNormalizedBitIndex(offset)};
+  return CheckBitStatus(byte_val, index);
+}
+
+bool GetBitValueSafe(const std::string& entry, uint32_t offset) {
+  return ((entry.size() * OFFSET_FACTOR) > (std::size_t)offset) ? GetBitValue(entry, offset)
+                                                                : false;
+}
+
+constexpr uint8_t TurnBitOn(uint8_t on, uint32_t offset) {
+  return on |= 1 << offset;
+}
+
+constexpr uint8_t TunBitOff(uint8_t on, uint32_t offset) {
+  return on &= ~(1 << offset);
+}
+
+bool SetBitValue(uint32_t offset, bool bit_value, std::string* entry) {
+  // we need to return the old value after setting the value for offset
+  const auto old_value{GetBitValue(*entry, offset)};  // save this as the return value
+  auto byte{GetByteValue(*entry, offset)};
+  std::bitset<8> bits{byte};
+  const auto bit_index{GetNormalizedBitIndex(offset)};
+  byte = bit_value ? TurnBitOn(byte, bit_index) : TunBitOff(byte, bit_index);
+  (*entry)[GetByteIndex(offset)] = byte;
+  return old_value;
+}
+
+// ------------------------------------------------------------------------- //
+// Helper functions to access the data or change it
+
+class OverrideValue {
+  EngineShard* shard_ = nullptr;
+  DbIndex index_ = 0;
+
+ public:
+  explicit OverrideValue(const OpArgs& args) : shard_{args.shard}, index_{args.db_ind} {
+  }
+
+  OpResult<bool> Set(std::string_view key, uint32_t offset, bool bit_value);
+};
+
+OpResult<bool> OverrideValue::Set(std::string_view key, uint32_t offset, bool bit_value) {
+  auto& db_slice = shard_->db_slice();
+  auto index = index_;
+
+  DCHECK(db_slice.IsDbValid(index_));
+
+  std::pair<PrimeIterator, bool> add_res;
+  try {
+    add_res = db_slice.AddOrFind(index_, key);
+  } catch (const std::bad_alloc&) {
+    return OpStatus::OUT_OF_MEMORY;
+  }
+  bool old_value = false;
+  PrimeIterator& it = add_res.first;
+  bool added = add_res.second;
+  auto UpdateBitMapValue = [&](std::string_view value) {
+    db_slice.PreUpdate(index, it);
+    it->second.SetString(value);
+    db_slice.PostUpdate(index, it, key, !added);
+  };
+
+  if (added) {  // this is a new entry in the "table"
+    std::string new_entry(GetByteIndex(offset) + 1, 0);
+    old_value = SetBitValue(offset, bit_value, &new_entry);
+    UpdateBitMapValue(new_entry);
+  } else {
+    if (it->second.ObjType() != OBJ_STRING) {
+      return OpStatus::WRONG_TYPE;
+    }
+    bool reset = false;
+    std::string existing_entry{GetString(shard_, it->second)};
+    if ((static_cast<decltype(offset)>(existing_entry.size()) * OFFSET_FACTOR) <=
```

Comment:
decltype is good for the template code. here it hurts readability.
what's wrong with:
`if (existing_entry.size() * OFFSET_FACTOR <= offset)` ? 
I personally find it more readable.

Reply:
I will change, the reason is that it produce a warning